lib/utils/ec2.go

```diff
@@ -84,3 +88,83 @@ func IsEC2NodeID(id string) bool {
 func NodeIDFromIID(iid *imds.InstanceIdentityDocument) string {
 	return iid.AccountID + "-" + iid.InstanceID
 }
+
+// InstanceMetadata is an interface for fetching information from EC2 instance
+// metadata.
+type InstanceMetadata interface {
+	// IsAvailable checks if instance metadata is available.
+	IsAvailable() bool
+	// GetTagKeys gets all of the EC2 tag keys.
+	GetTagKeys() ([]string, error)
+	// GetTagValue gets the value for a specified tag key.
+	GetTagValue(key string) (string, error)
+}
+
+// InstanceMetadataClient is a wrapper for an imds.Client.
+type InstanceMetadataClient struct {
+	c   *imds.Client
+	ctx context.Context
```

Comment:
I don't think it makes sense to store the context with the client, and the docs recommend against it https://pkg.go.dev/context
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.